pkg/protocol/chains.go

```diff
@@ -0,0 +1,180 @@
+		HeaviestClaimedCandidate:  reactive.NewVariable[*Chain](),
+		HeaviestAttestedCandidate: reactive.NewVariable[*Chain](),
+		HeaviestVerifiedCandidate: reactive.NewVariable[*Chain](),
+		LatestSeenSlot:            reactive.NewVariable[iotago.SlotIndex](increasing[iotago.SlotIndex]),
+		protocol:                  protocol,
+	}
+
+	shutdown := lo.Batch(
+		c.initLogger(protocol.NewChildLogger("Chains")),
+		c.initChainSwitching(protocol.Options.ChainSwitchingThreshold),
+
+		protocol.Constructed.WithNonEmptyValue(func(_ bool) (shutdown func()) {
+			return c.deriveLatestSeenSlot(protocol)
+		}),
+	)
+
+	protocol.Shutdown.OnTrigger(shutdown)
+
+	return c
+}
+
+// WithInitializedEngines is a reactive selector that executes the given callback for each managed chain that
+// initialized its engine.
+func (c *Chains) WithInitializedEngines(callback func(chain *Chain, engine *engine.Engine) (shutdown func())) (shutdown func()) {
+	return c.WithElements(func(chain *Chain) (shutdown func()) {
+		return chain.WithInitializedEngine(func(engine *engine.Engine) (shutdown func()) {
+			return callback(chain, engine)
+		})
+	})
+}
+
+// initLogger initializes the logger for this component.
+func (c *Chains) initLogger(logger log.Logger, shutdownLogger func()) (shutdown func()) {
+	c.Logger = logger
+
+	return lo.Batch(
+		c.Main.LogUpdates(c, log.LevelTrace, "Main", (*Chain).LogName),
+		c.HeaviestClaimedCandidate.LogUpdates(c, log.LevelTrace, "HeaviestClaimedCandidate", (*Chain).LogName),
+		c.HeaviestAttestedCandidate.LogUpdates(c, log.LevelTrace, "HeaviestAttestedCandidate", (*Chain).LogName),
+		c.HeaviestVerifiedCandidate.LogUpdates(c, log.LevelTrace, "HeaviestVerifiedCandidate", (*Chain).LogName),
+
+		shutdownLogger,
+	)
+}
+
+// initChainSwitching initializes the chain switching logic.
+func (c *Chains) initChainSwitching(chainSwitchingThreshold iotago.SlotIndex) (shutdown func()) {
+	mainChain := c.newChain()
+	mainChain.StartEngine.Set(true)
+
+	c.Main.Set(mainChain)
+
+	return lo.Batch(
+		c.HeaviestClaimedCandidate.WithNonEmptyValue(func(heaviestClaimedCandidate *Chain) (shutdown func()) {
+			return heaviestClaimedCandidate.RequestAttestations.ToggleValue(true)
+		}),
+
+		c.HeaviestAttestedCandidate.WithNonEmptyValue(func(heaviestAttestedCandidate *Chain) (shutdown func()) {
+			return heaviestAttestedCandidate.StartEngine.ToggleValue(true)
+		}),
+
+		c.HeaviestVerifiedCandidate.WithNonEmptyValue(func(heaviestVerifiedCandidate *Chain) (shutdown func()) {
+			// only switch to the heaviest chain if the latest produced commitment is enough slots away from the forking point.
+			chainSwitchingCondition := func(_ *Commitment, latestProducedCommitment *Commitment) bool {
+				forkingPoint := heaviestVerifiedCandidate.ForkingPoint.Get()
+
+				return forkingPoint != nil && latestProducedCommitment != nil && (latestProducedCommitment.ID().Slot()-forkingPoint.ID().Slot()) > chainSwitchingThreshold
```

Comment:
This is not exactly like the fork choice rule was implemented before. Right now you're checking whether the candidate chain is heavier for x slots when producing the commitments ourselves from that candidate engine. However, you instantiate a candidate engine immediately after another attested chain becomes heavier. Before, a candidate engine would only be instantiated when x attested slots in a row were heavier. So I think we should add this check additionally to `HeaviestAttestedCandidate` in order to make sure that we're not tricked into spawning a new engine

Reply:
I fixed it